File: assets/js/engine/branding.js
/**
 * Planes y resolución de branding — puerto JS de `backend/app/core/{config,branding}.py`.
 *
 * El gating vive en UN solo lugar por motor. Hoy el navegador siempre corre en
 * plan `free` (el modo personalizado se desbloquea con licencia contra el backend),
 * pero la tabla completa está aquí para que activar el cobro sea cambiar el plan
 * que devuelve `resolvePlanKey`, no reescribir la app.
 *
 * @module engine/branding
 */

export const WATERMARK_TEXT = 'CronoWeb.com';

export const PLAN_FEATURES = Object.freeze({
  free: {
    key: 'free',
    label: 'Modo simple',
    price_mxn_year: 0,
    custom_branding: false,
    remove_watermark: false,
    max_groups: 12,
    max_teachers: 40,
    max_time_budget_seconds: 10,
    persistence: false,
    api_access: false,
  },
  school: {
    key: 'school',
    label: 'Escuela',
    price_mxn_year: 1800,
    custom_branding: true,
    remove_watermark: false,
    max_groups: 60,
    max_teachers: 200,
    max_time_budget_seconds: 60,
    persistence: true,
    api_access: false,
  },
  zone: {
    key: 'zone',
    label: 'Zona escolar',
    price_mxn_year: 12000,
    custom_branding: true,
    remove_watermark: false,
    max_groups: 600,
    max_teachers: 2000,
    max_time_budget_seconds: 120,
    persistence: true,
    api_access: true,
  },
});

export const planFeatures = (plan) => PLAN_FEATURES[String(plan || 'free').toLowerCase()] || PLAN_FEATURES.free;

/**
 * Resuelve el branding efectivo. Regla de negocio: la marca de agua CronoWeb.com
 * nunca desaparece del PNG; en modo personalizado baja a crédito discreto.
 * @returns {{resolved:object, conflicts:object[]}}
 */
export function resolveBranding(branding, plan) {
  const features = planFeatures(plan);
  const conflicts = [];
  const wantsCustom = branding?.mode === 'custom';
  const allowCustom = wantsCustom && features.custom_branding;
  const downgraded = wantsCustom && !features.custom_branding;

  if (downgraded) {
    conflicts.push({
      severity: 'info',
      code: 'BRANDING_DOWNGRADED',
      message: `El plan «${features.label}» no incluye personalización; el horario se generó en ` +
        'modo simple con marca de agua CronoWeb.com.',
      group_id: null, subject_id: null, teacher_id: null, missing_hours: null,
    });
  }

  const resolved = allowCustom
    ? {
      mode: 'custom',
      plan: features.key,
      school_name: branding.school_name || null,
      logo_data_url: branding.logo_data_url || null,
      primary_color: branding.primary_color || '#0f766e',
      cycle_label: branding.cycle_label || null,
      footer_note: branding.footer_note || null,
      show_logo: Boolean(branding.logo_data_url),
      watermark_text: WATERMARK_TEXT,
      watermark_required: !features.remove_watermark,
      downgraded: false,
    }
    : {
      // Modo simple: se descarta todo dato de identidad de la escuela.
      mode: 'simple',
      plan: features.key,
      school_name: null,
      logo_data_url: null,
      primary_color: '#0f766e',
      cycle_label: branding?.cycle_label || null,
      footer_note: null,
      show_logo: false,
      watermark_text: WATERMARK_TEXT,
      watermark_required: true,
      downgraded,
    };

  return { resolved, conflicts };
}

File: backend/app/core/branding.py
"""
Resolución de branding según el plan.

Regla de negocio: la marca de agua **CronoWeb.com** nunca desaparece del PNG.
En modo simple es el sello grande; en modo personalizado se degrada a crédito
discreto al pie. Si un plan sin `custom_branding` pide modo personalizado, se
degrada silenciosamente a simple y se devuelve un conflicto `info` (no un error:
el horario sí se generó).
"""

from __future__ import annotations

from typing import List, Tuple

from ..models.schemas import Branding, BrandingMode, Conflict, ResolvedBranding, Severity
from .config import plan_features

WATERMARK_TEXT = "CronoWeb.com"


def resolve_branding(branding: Branding, plan: str) -> Tuple[ResolvedBranding, List[Conflict]]:
    features = plan_features(plan)
    conflicts: List[Conflict] = []

    wants_custom = branding.mode is BrandingMode.CUSTOM
    allowed_custom = wants_custom and features.custom_branding
    downgraded = wants_custom and not features.custom_branding

    if downgraded:
        conflicts.append(
            Conflict(
                severity=Severity.INFO,
                code="BRANDING_DOWNGRADED",
                message=(
                    f"El plan «{features.label}» no incluye personalización; "
                    "el horario se generó en modo simple con marca de agua CronoWeb.com."
                ),
            )
        )

    if allowed_custom:
        resolved = ResolvedBranding(
            mode=BrandingMode.CUSTOM,
            plan=features.key,
            school_name=branding.school_name,
            logo_data_url=branding.logo_data_url,
            primary_color=branding.primary_color,
            cycle_label=branding.cycle_label,
            footer_note=branding.footer_note,
            show_logo=bool(branding.logo_data_url),
            watermark_text=WATERMARK_TEXT,
            # En plan de pago la marca sigue presente, pero como crédito al pie.
            watermark_required=not features.remove_watermark,
            downgraded=False,
        )
    else:
        # Modo simple: se descarta todo dato de identidad de la escuela.
        resolved = ResolvedBranding(
            mode=BrandingMode.SIMPLE,
            plan=features.key,
            school_name=None,
            logo_data_url=None,
            primary_color="#14417c",
            cycle_label=branding.cycle_label,
            footer_note=None,
            show_logo=False,
            watermark_text=WATERMARK_TEXT,
            watermark_required=True,
            downgraded=downgraded,
        )

    return resolved, conflicts
